rfc-173-testing-our-work.md

```diff
@@ -0,0 +1,68 @@
+
+## Problem
+
+- Testing pattern divergence - the subject of writing tests is a broad one and
+  there are naturally varying opinions on what patterns to follow in any
+  situation. The patterns we use have been slowing diverging, we think this is
+  largely a natural side affect of different team compositions and their views
+  and lack of explicit standards to follow
+- Tooling divergence - we work with the philosophy and freedom of using the best
+  tool for the job. It is okay to diverge when there is good reason but we could
+  benefit from using like-for-like tools through defining some defaults. The
+  impact of using lots of different tooling is that projects are harder for us
+  to swap between and are harder for us to support
+- Team growth - the development team is growing and we often invite contractors
+  to represent us. Having an agreed set of principles will help all new joiners
+  build an understanding around what we expect from one another when writing
+  software
+- Stack variety - dxw are increasingly working with different stacks. These
+  often require us to use different testing tools. Having a way to record our
+  preferences as we grow will make it easier to contribute or work on those
+  stacks again in future
+
+## Proposal
+
+The topic of testing code has many facets. The aim of this proposal is to
+capture some of the higher level principles we wish to follow, seperate to how
+we achieve them. We propose to document those finer details in our
+[standards for testing](https://docs.google.com/document/d/1pSZkrviFAiR-0U-PGixOyC2MfiBdCIWIN66YyLa23nc/edit#heading=h.owf78h5xtccv)
+that may change more frequently. These standards may eventually become a guide
+in [the dxw Playbook](https://github.com/dxw/playbook).
+
+- We MUST write automated tests for all the code we write that will (or could)
+  serve people in live environments – this includes internal tools for dxw staff
+  and tooling that provisions real infrastructure
+- We MUST write appropriate test coverage for the code we write, inline with our
+  standards for testing. This includes using a combination of feature, request
+  and unit tests. It will also involve setting testing boundaries between
+  systems and third-party software
+- We MUST configure our continuous integration tools to automatically run as
+  deployment pipelines. It MUST prevent deployments completing should any test
+  fail
+- We MUST set up new projects to require 100% code coverage
+
+- We SHOULD write our tests in a consistent and readable way that follows our
+  conventions and patterns set out in our standards for testing
+- We SHOULD use consistent tooling for testing where possible
+- We SHOULD write architecture decision records (ADR) within each project that
+  chooses to diverge from our tools or standards for testing diverge from dxw
+  testing standards
+- We SHOULD aim to increase the minimum test coverage on projects that don't
+  already require 100% as high as we can
+
```

Comment:
Is it worth adding a couple of `SHOULD`s here about browser testing e.g if there's client side javascript or for cross-browser testing as we've got in the standards for testing document?

Reply:
Added browser and OS as a SHOULD 👍 

Testing without javascript is possibly a bit tricker because it's client dependant. We definitely should with apps following the service standard otherwise it'll fail but we can be involved in building JS apps where that standard doesn't apply? I'll add it with a caveat.